Retry scheduling on StorageClass events

cofyc opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #56163

**Special notes for your reviewer**:

I have taken over #60006.
It's hard to test in e2e, because we cannot know reschedule of pod is triggered by which event (periodically service/node events will move pods to active queue too). ~~I'll add integration tests for this functionality after [this PR](https://github.com/kubernetes/kubernetes/pull/65296) get merged.~~ (already added)

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

cofyc commented:
hi, @msau42 @bsalamat 

I've updated PR to add integration tests. Please help reviewing it!

Summary of changes:

pkg/scheduler/factory/factory.go:

- move pods to active queue on a WaitForFirstConsumer storage class add
- invalidate CheckVolumeBindingPred/NoVolumeZoneConflictPred predicates cache when a WaitForFirstConsumer storage class is deleted

test/integration/scheduler/util.go:

- add `resyncPeriod` for initTestSchedulerWithOptions to specify resync period in scheduler
- initTestScheduler/initTestDisablePreemption use default `time.Second` to keep backward compatbility
  - kube-scheduler disabled resync period, why using 1 time.Second resync period in integration tests?

test/integration/scheduler/volume_binding_test.go:

- add `features` for setupCluster to configure features and restore when done
- add `FakeVolumePlugin` in setupCluster to mock dynamic provisioning
- use `initTestSchedulerWithOptions` to initialize scheduler with 0 resync period
  - previously, test scheduler use 1 second resync period which will move pods into queue every resync (trigger objects update events even if they are not updated)
- add `TestVolumeBindingRescheduling` to test scheduler will retry scheduling on storage class add

About integration tests for CheckVolumeBindingPred/NoVolumeZoneConflictPred invalidation on storage deletion, how about adding them in a separate PR (maybe after https://github.com/kubernetes/kubernetes/pull/62226 get merged)?

msau42 commented on test/integration/scheduler/volume_binding_test.go:
```diff
@@ -361,37 +459,61 @@ func TestPVAffinityConflict(t *testing.T) {
 	controllerCh := make(chan struct{})
 
-	context := initTestScheduler(t, initTestMaster(t, nsName, nil), controllerCh, false, nil)
+	context := initTestSchedulerWithOptions(t, initTestMaster(t, nsName, nil), controllerCh, false, nil, false, 0)
```
the resync period of 1 second actually uncovered a bug in the volume binder.  It may be good to make resync configurable here too

cofyc replied:
Like this? https://github.com/kubernetes/kubernetes/pull/65616/commits/0c25d57daccf85363d7aec98f39a21bb9d9aa8d8

msau42 commented on pkg/scheduler/factory/factory.go:
```diff
@@ -562,6 +570,59 @@ func (c *configFactory) invalidatePredicatesForPvcUpdate(old, new *v1.Persistent
+	}
+}
+
+func (c *configFactory) invalidatePredicatesForStorageClass(sc *storagev1.StorageClass) {
```
Please add this scenario to the list of test cases we need to add for predicate invalidation

cofyc replied:
Added in https://github.com/kubernetes/kubernetes/issues/66148.

msau42 commented on test/integration/scheduler/volume_binding_test.go:
```diff
@@ -246,9 +255,94 @@ func TestVolumeBinding(t *testing.T) {
+		trigger    func(config *testConfig)
+		shouldFail bool
+	}{
+		"reschedule on WaitForFirstConsumer storage class added": {
```
For static binding, you could also have this scenario where:
- create pvc, there are no pvs available
- create storageclass
- create pvs

cofyc replied:
@msau42  updated in https://github.com/kubernetes/kubernetes/pull/65616/commits/8eb7a3b945f4db77a938f310921ee9c80fff169f

In testing this scenario, I found we need to move pods into active queue on PV add & update:

- on PV add

make sure unbounded PVs created for static provisioning and delay binding storage class will reschedule stuck pods 
- on PV update

IIUC, `Scheduler.bindVolumesWorker` needs to send back pod to scheduler always, so it fails with error in all cases, but [error handler only puts pod back to unschedule queue on error](https://github.com/kubernetes/kubernetes/blob/v1.12.0-alpha.0/pkg/scheduler/factory/factory.go#L1340). If `Scheduler.bindVolumesWorker` failed to update PV due to conflicts when PV is modified by others, PVC will not be bound to volume by PV controller, so [no PVC update event to move pods to active queue](https://github.com/kubernetes/kubernetes/blob/v1.12.0-alpha.0/pkg/scheduler/factory/factory.go#L486). In this case, we need to move pods to active queue on PV update event.

cofyc commented:
/test pull-kubernetes-e2e-kops-aws

cofyc commented on test/integration/scheduler/volume_binding_test.go:
```diff
@@ -342,7 +342,7 @@ func TestVolumeBindingStress(t *testing.T) {
 		"PersistentLocalVolumes": true,
 	}
-	config := setupCluster(t, "volume-binding-stress", 1, features)
+	config := setupCluster(t, "volume-binding-stress", 1, features, 0)
```
Added in https://github.com/kubernetes/kubernetes/pull/65616/commits/31f666c535748b94491bd834ae1a3b809bc4e6f2.

msau42 commented:
/lgtm

msau42 commented:
ah, can you squash the commits a bit? maybe one for the code, and one for the test

Commit:
Retry scheduling on various events.

Commit:
Test rescheduling on various events.
- Add resyncPeriod parameter for setupCluster to make resync period of
scheduler configurable.
- Add test case for static provisioning and delay binding storage class.
Move pods into active queue on PV add/update events.
- Add a stress test with scheduler resync to detect possible race
conditions.

cofyc commented:
@msau42 done
- https://github.com/kubernetes/kubernetes/pull/65616/commits/8f0373792f8b7941f4742abc0a048b43a8365ca6 for code
- https://github.com/kubernetes/kubernetes/pull/65616/commits/c868b5bc881ed2fbb453abf817ec709f57aba33c for test

msau42 commented:
/lgtm

cofyc commented:
/test pull-kubernetes-e2e-kops-aws

cofyc commented:
/test pull-kubernetes-e2e-kops-aws

cofyc commented:
/assign @bsalamat 
for scheduler approval

cofyc commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce

bsalamat reviewed (approved):
/approve
